pennylane/tracker.py

```diff
@@ -0,0 +1,195 @@
+
+    Args:
+        dev (Device): A PennyLane compatible device
+        callback=None (callable or None): A function of the keywords ``totals``,
+            ``history`` and ``latest``.  Run on each ``record`` call with current values of
+            the corresponding attributes.
+        persistent=False (bool): Whether to reset stored information upon
+            entering a runtime context.
+
+
+    **Example**
+
+    Using a ``with`` statement to toggle the active mode, we can see the number of executions
+    and shots used to calculate a parameter-shift derivative.
+
+    .. code-block:: python
+
+        dev = qml.device('default.qubit', wires=1, shots=100)
+
+        @qml.qnode(dev, diff_method="parameter-shift")
+        def circuit(x):
+            qml.RX(x, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        x = np.array(0.1)
+
+        with qml.Tracker(dev) as tracker:
+            qml.grad(circuit)(x)
+
+    You can then access the tabulated information through ``totals``, ``history``, and ``latest``:
+
+    >>> tracker.totals
+    {'executions': 3, 'shots': 300, 'batches': 1, 'batch_len': 2}
+    >>> tracker.history
+    {'executions': [1, 1, 1],
+     'shots': [100, 100, 100],
+     'batches': [1],
+     'batch_len': [2]}
+    >>> tracker.latest
+    {'batches': 1, 'batch_len': 2}
+
+    We can see that calculating the gradient of ``circuit`` takes three total evaluations: one
+    forward pass and one batch of length two for the derivative of ``qml.RX``.
+
+    .. UsageDetails::
+
+        .. note::
+            With backpropagation, this function should take ``qnode.device``
+            instead of the device used to create the QNode.
+
+    Users can pass a custom callback function to the ``callback`` keyword. This
+    function is run each time the ``record()`` method is called, which occurs near
+    the end of a device's ``execute`` and ``batch_execute`` methods. Using ``print``
+    or logging, users can monitor completion during a long set of jobs.
+
+    The function passed must accept ``totals``, ``history``, and ``latest`` as
+    keyword arguments. The dictionary ``latest`` will contain different keywords based on whether
+    whether ``execute`` or ``batch_execute`` last performed an update.
+
+    >>> def shots_info(totals, history, latest):
+    ...     if 'shots' in latest.keys():
+    ...         print("Total shots: ", totals['shots'])
+    >>> with qml.Tracker(circuit.device, callback=shots_info) as tracker:
+    ...     qml.grad(circuit)(0.1)
+    Total shots:  100
+    Total shots:  200
+    Total shots:  300
+
+    By specifying ``persistent=False``, you can reuse the same tracker across
+    multiple contexts.
+
+    >>> with qml.Tracker(circuit.device, persistent=False) as tracker:
+    ...     circuit(0.1)
+    >>> with tracker:
+    ...     circuit(0.2)
+    >>> tracker.totals['executions']
+    2
```

Comment:
@albi3ro this all needs to be indented to properly display inside the Usage Details dropdown!